rfcs/0049-flakes.md

```diff
@@ -0,0 +1,709 @@
+
+Most `nix` subcommands work on a list of arguments called
+"installables" for lack of a better word. For example,
+
+    # nix run nixpkgs:hello dwarffs:dwarffs
+
+takes two flake-based installables. The general form is:
+
+    <flake-ref>(:<attr-path>)?
+
+Examples of installables:
+
+* `nixpkgs:packages.hello`
+* `nixpkgs:hello` - short for `nixpkgs:packages.hello`
+* `nixpkgs/release-19.03:hello` - overrides the Git branch to use
+* `github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082:hello` -
+  specifies the exact Git revision to use
+* `dwarffs` - short for `dwarffs:defaultPackage`
+* `nix:hydraJobs.build.x86_64-darwin`
+* `.:hydraJobs.build.x86_64-darwin` - refers to the flake in the
+  current directory (which can be a dirty Git tree)
+* `.` - short for `.:defaultPackage`
+
+If no argument is given, the default is `.`; thus,
+
+    # nix build
+
+is equivalent to
+
+    # nix build .:defaultPackage
+
+For backwards compatibility, it's possible to use non-flake Nix
+expressions using `-f`, e.g. `nix build -f foo.nix foo.bar`.
+
+## Lock files and reproducible evaluation
+
+Inputs specified in `flake.nix` are typically "unlocked": they are
+flake references that don't specify an exact revision (e.g. `nixpkgs`
+rather than
+`github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082`). To
+ensure reproducibility, Nix will automatically generate and use a
+*lock file* called `flake.lock` in the flake's directory. The lock
+file contains a tree of mappings from the flake references specified
+in `flake.nix` to direct flake references that contain revisions.
+
+For example, if `flake.nix` contains
+```
+inputs =
+  [ "nixpkgs"
+    github:edolstra/import-cargo
+  ];
+```
+then the resulting lock file might be:
+```
+{
+    "version": 2,
+    "inputs": {
+        "github:edolstra/import-cargo": {
+            "id": "import-cargo",
+            "inputs": {},
+            "narHash": "sha256-mxwKMDFOrhjrBQhIWwwm8mmEugyx/oVlvBH1CKxchlw=",
+            "uri": "github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e"
+        },
+        "nixpkgs": {
+            "id": "nixpkgs",
+            "inputs": {},
+            "narHash": "sha256-p7UqhvhwS5MZfqUbLbFm+nfG/SMJrgpNXxWpRMFif8c=",
+            "uri": "github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082"
+        }
+    }
+}
+```
+
+Thus, when we build this flake, the input `nixpkgs` is mapped to
+`github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e`. Nix
+will also check that the content hash of the input is equal to the one
+recorded in the lock file. This check is superfluous for Git
+repositories (since the commit hash serves a similar purpose), but for
+GitHub archives, we cannot directly check that the contents match the
+commit hash.
+
+Note that lock files are only used at top-level: the `flake.lock` files
+in dependencies (if they exist) are ignored.
+
+Nix automatically creates a `flake.lock` file when you build a local
```

Comment:
Rust’s `cargo` automatically creates a lock file, and it leads to a nice user experience, because locking the version you are working with is always what you want.

Reply:
My experience is that if the lockfile is not generated by default, developers forget to commit the lockfile. Which then introduces all sort of reproducibility problems. When the lockfile is generated automatically it becomes easy to compare changes on updates.

Reply:
Lockfiles should not be commited anyway as this is effectively no different from [vendor bundling](https://fedoraproject.org/wiki/Bundled_Libraries).

I agree that generation of lockfiles should be opt-in.

Reply:
Unlike bundling, the lockfiles don't prevent people from using a different dependency set. What it does it provide a well know working set that can be used to build the project reproducibly.